src/Emulator/Peripherals/Peripherals/UART/STM32_UART.cs

```diff
@@ -164,11 +184,14 @@ private void DefineRegisters()
 
         private void Update()
         {
-            IRQ.Set(
-                (receiverNotEmptyInterruptEnabled.Value && readFifoNotEmpty.Value) ||
-                (transmitDataRegisterEmptyInterruptEnabled.Value) || // TXE is assumed to be true
-                (transmissionCompleteInterruptEnabled.Value && transmissionComplete.Value)
-            );
+            machine.ClockSource.ExecuteInLock(() =>
```

Comment:
FYI:

Here is a possible dead lock. I don't like my fix but it worked for me.

_machine thread_ is inside BaseClockSource (locked) and tries to lock the NVIC.

_\_Reader\_Thread_ updates the NVIC (locked) and then tries to lock the BaseClockSource which is locked by _machine thread_.


![Screenshot_20210114_142013](https://user-images.githubusercontent.com/3637911/104606135-ac827c00-567f-11eb-9955-fb9da2255f72.png)

Reply:
@Viatorus  Sorry, How do you generate this figure? Could you please tell me the name of the tool?

Reply:
It's a debug tool from Rider: https://www.jetbrains.com/help/rider/Debugging_Multithreaded_Applications.html#parallel-stacks